File: FuturesQuant/core/indicator/mlmi.py
"""
Machine Learning Momentum Index (MLMI) implementation
Based on Zeiierman's Pine Script implementation
"""

import numpy as np
import pandas as pd
from typing import Dict, Any, List
from .base import Indicator

class MLMI(Indicator):
    def __init__(self, 
                 num_neighbors: int = 200,
                 momentum_window: int = 20):
        """
        Initialize the MLMI indicator.
        
        Args:
            num_neighbors (int): Number of neighbors for k-NN prediction
            momentum_window (int): Window for momentum calculation
        """
        super().__init__("MLMI", {
            "num_neighbors": num_neighbors,
            "momentum_window": momentum_window
        })
        
        self.num_neighbors = num_neighbors
        self.momentum_window = momentum_window
        
        # Initialize data storage
        self.parameter1 = []  # RSI slow
        self.parameter2 = []  # RSI quick
        self.price_array = []
        self.result_array = []
        
    def calculate(self, data: pd.DataFrame) -> Dict[str, Any]:
        """
        Calculate the MLMI indicator values.
        
        Args:
            data (pd.DataFrame): DataFrame containing price data
            
        Returns:
            Dict containing:
            - prediction: Current MLMI prediction
            - prediction_ma: WMA of prediction
            - upper: Upper channel
            - lower: Lower channel
            - upper_: Upper channel with standard deviation
            - lower_: Lower channel with standard deviation
            - signal: Trading signal based on crossovers
        """
        if len(data) < self.momentum_window:
            return {
                "prediction": 0,
                "prediction_ma": 0,
                "upper": 0,
                "lower": 0,
                "upper_": 0,
                "lower_": 0,
                "signal": None
            }
            
        # Calculate moving averages and RSI
        ma_quick = self.calculate_wma(data['close'], 5)
        ma_slow = self.calculate_wma(data['close'], 20)
        
        rsi_quick = self.calculate_wma(self.calculate_rsi(data['close'], 5), self.momentum_window)
        rsi_slow = self.calculate_wma(self.calculate_rsi(data['close'], 20), self.momentum_window)
        
        # Check for crossovers
        pos = self.crossover(ma_quick, ma_slow)
        neg = self.crossunder(ma_quick, ma_slow)
        
        # Store previous trade data if crossover occurs
        if pos or neg:
            self.store_previous_trade(rsi_slow, rsi_quick)
            
        # Calculate prediction
        prediction = self.knn_predict(rsi_slow, rsi_quick)
        prediction_ma = self.calculate_wma(pd.Series(prediction), 20)
        
        # Calculate channels
        upper = np.max(prediction)
        lower = np.min(prediction)
        std_dev = self.calculate_ema(self.calculate_std(prediction, 20), 20)
        
        upper_ = upper - std_dev
        lower_ = lower + std_dev
        
        # Generate signal based on crossovers
        signal = None
        if prediction > upper_:
            signal = 'SELL'
        elif prediction < lower_:
            signal = 'BUY'
            
        return {
            "prediction": prediction,
            "prediction_ma": prediction_ma,
            "upper": upper,
            "lower": lower,
            "upper_": upper_,
            "lower_": lower_,
            "signal": signal
        }
        
    def store_previous_trade(self, p1: float, p2: float):
        """Store previous trade data."""
        if len(self.parameter1) > 0:
            self.parameter1.append(self.parameter1[-1])
            self.parameter2.append(self.parameter2[-1])
            self.price_array.append(self.price_array[-1])
            self.result_array.append(1 if self.price_array[-1] >= self.price_array[-2] else -1)
            
        self.parameter1.append(p1)
        self.parameter2.append(p2)
        self.price_array.append(self.price_array[-1] if len(self.price_array) > 0 else 0)
        
    def knn_predict(self, p1: float, p2: float) -> float:
        """Make prediction using k-NN algorithm."""
        if len(self.parameter1) < 2:
            return 0
            
        # Calculate distances
        distances = []
        for i in range(len(self.parameter1)):
            dist = np.sqrt(
                np.power(p1 - self.parameter1[i], 2) + 
                np.power(p2 - self.parameter2[i], 2)
            )
            distances.append(dist)
            
        # Get k nearest neighbors
        k = min(self.num_neighbors, len(distances))
        nearest_indices = np.argsort(distances)[:k]
        neighbors = [self.result_array[i] for i in nearest_indices]
        
        return np.sum(neighbors)
        
    def calculate_wma(self, data: pd.Series, period: int) -> pd.Series:
        """Calculate Weighted Moving Average."""
        weights = np.arange(1, period + 1)
        return data.rolling(window=period).apply(
            lambda x: np.sum(weights * x) / weights.sum()
        )
        
    def calculate_rsi(self, data: pd.Series, period: int) -> pd.Series:
        """Calculate Relative Strength Index."""
        delta = data.diff()
        gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
        loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
        rs = gain / loss
        return 100 - (100 / (1 + rs))
        
    def calculate_ema(self, data: pd.Series, period: int) -> pd.Series:
        """Calculate Exponential Moving Average."""
        return data.ewm(span=period, adjust=False).mean()
        
    def calculate_std(self, data: pd.Series, period: int) -> pd.Series:
        """Calculate Standard Deviation."""
        return data.rolling(window=period).std()
        
    def crossover(self, series1: pd.Series, series2: pd.Series) -> bool:
        """Check for crossover."""
        return series1.iloc[-1] > series2.iloc[-1] and series1.iloc[-2] <= series2.iloc[-2]
        
    def crossunder(self, series1: pd.Series, series2: pd.Series) -> bool:
        """Check for crossunder."""
        return series1.iloc[-1] < series2.iloc[-1] and series1.iloc[-2] >= series2.iloc[-2] 

File: FuturesQuant/src/strategy/mlmi.js
/**
 * Machine Learning Momentum Index (MLMI) implementation
 * Based on Zeiierman's Pine Script implementation
 */

class MLMI {
    constructor(options = {}) {
        // Default settings
        this.numNeighbors = options.numNeighbors || 200;
        this.momentumWindow = options.momentumWindow || 20;
        
        // Initialize data storage
        this.parameter1 = [];  // RSI slow
        this.parameter2 = [];  // RSI quick
        this.priceArray = [];
        this.resultArray = [];
    }
    
    update(candleData) {
        // Add candle data to history
        this.data.push(candleData);
        
        // Keep only recent history
        if (this.data.length > this.momentumWindow) {
            this.data.shift();
        }
        
        if (this.data.length < this.momentumWindow) {
            return {
                prediction: 0,
                prediction_ma: 0,
                upper: 0,
                lower: 0,
                upper_: 0,
                lower_: 0,
                signal: null
            };
        }
        
        // Calculate moving averages and RSI
        const maQuick = this.calculateWMA(this.data.map(d => d.close), 5);
        const maSlow = this.calculateWMA(this.data.map(d => d.close), 20);
        
        const rsiQuick = this.calculateWMA(this.calculateRSI(this.data.map(d => d.close), 5), this.momentumWindow);
        const rsiSlow = this.calculateWMA(this.calculateRSI(this.data.map(d => d.close), 20), this.momentumWindow);
        
        // Check for crossovers
        const pos = this.crossover(maQuick, maSlow);
        const neg = this.crossunder(maQuick, maSlow);
        
        // Store previous trade data if crossover occurs
        if (pos || neg) {
            this.storePreviousTrade(rsiSlow, rsiQuick);
        }
        
        // Calculate prediction
        const prediction = this.knnPredict(rsiSlow, rsiQuick);
        const predictionMA = this.calculateWMA([prediction], 20);
        
        // Calculate channels
        const upper = Math.max(...this.resultArray);
        const lower = Math.min(...this.resultArray);
        const stdDev = this.calculateEMA(this.calculateStd([prediction], 20), 20);
        
        const upper_ = upper - stdDev;
        const lower_ = lower + stdDev;
        
        // Generate signal based on crossovers
        let signal = null;
        if (prediction > upper_) {
            signal = 'SELL';
        } else if (prediction < lower_) {
            signal = 'BUY';
        }
        
        return {
            prediction,
            prediction_ma: predictionMA,
            upper,
            lower,
            upper_,
            lower_,
            signal
        };
    }
    
    storePreviousTrade(p1, p2) {
        if (this.parameter1.length > 0) {
            this.parameter1.push(this.parameter1[this.parameter1.length - 1]);
            this.parameter2.push(this.parameter2[this.parameter2.length - 1]);
            this.priceArray.push(this.priceArray[this.priceArray.length - 1]);
            this.resultArray.push(
                this.priceArray[this.priceArray.length - 1] >= this.priceArray[this.priceArray.length - 2] ? 1 : -1
            );
        }
        
        this.parameter1.push(p1);
        this.parameter2.push(p2);
        this.priceArray.push(this.priceArray.length > 0 ? this.priceArray[this.priceArray.length - 1] : 0);
    }
    
    knnPredict(p1, p2) {
        if (this.parameter1.length < 2) {
            return 0;
        }
        
        // Calculate distances
        const distances = this.parameter1.map((param1, i) => {
            const param2 = this.parameter2[i];
            return Math.sqrt(
                Math.pow(p1 - param1, 2) + 
                Math.pow(p2 - param2, 2)
            );
        });
        
        // Get k nearest neighbors
        const k = Math.min(this.numNeighbors, distances.length);
        const sortedIndices = distances
            .map((dist, i) => ({ dist, i }))
            .sort((a, b) => a.dist - b.dist)
            .slice(0, k)
            .map(item => item.i);
            
        const neighbors = sortedIndices.map(i => this.resultArray[i]);
        
        return neighbors.reduce((sum, val) => sum + val, 0);
    }
    
    calculateWMA(data, period) {
        const weights = Array.from({ length: period }, (_, i) => i + 1);
        const sum = weights.reduce((a, b) => a + b, 0);
        
        let result = 0;
        for (let i = 0; i < period; i++) {
            result += data[data.length - 1 - i] * weights[i];
        }
        
        return result / sum;
    }
    
    calculateRSI(data, period) {
        const deltas = [];
        for (let i = 1; i < data.length; i++) {
            deltas.push(data[i] - data[i - 1]);
        }
        
        const gains = deltas.map(d => d > 0 ? d : 0);
        const losses = deltas.map(d => d < 0 ? -d : 0);
        
        const avgGain = this.calculateWMA(gains, period);
        const avgLoss = this.calculateWMA(losses, period);
        
        const rs = avgGain / avgLoss;
        return 100 - (100 / (1 + rs));
    }
    
    calculateEMA(data, period) {
        const k = 2 / (period + 1);
        let ema = data[0];
        
        for (let i = 1; i < data.length; i++) {
            ema = data[i] * k + ema * (1 - k);
        }
        
        return ema;
    }
    
    calculateStd(data, period) {
        const mean = this.calculateMean(data);
        const squaredDiffs = data.map(val => Math.pow(val - mean, 2));
        return Math.sqrt(this.calculateMean(squaredDiffs));
    }
    
    calculateMean(values) {
        return values.reduce((sum, val) => sum + val, 0) / values.length;
    }
    
    crossover(series1, series2) {
        return series1[series1.length - 1] > series2[series2.length - 1] &&
               series1[series1.length - 2] <= series2[series2.length - 2];
    }
    
    crossunder(series1, series2) {
        return series1[series1.length - 1] < series2[series2.length - 1] &&
               series1[series1.length - 2] >= series2[series2.length - 2];
    }
}

module.exports = MLMI; 